Mathlib/Util/Imports.lean

```diff
@@ -0,0 +1,113 @@
+
+/--
+Find the imports of a given module.
+-/
+def importsOf (env : Environment) (n : Name) : Array Name :=
+  if n = env.header.mainModule then
+    env.header.imports.map Import.module
+  else match env.getModuleIdx? n with
+    | .some idx => env.header.moduleData[idx.toNat]!.imports.map Import.module |>.erase `Init
+    | .none => #[]
+
+/--
+Construct the import graph of the current file.
+-/
+partial def importGraph (env : Environment) : NameMap (Array Name) :=
+  let main := env.header.mainModule
+  let imports := env.header.imports.map Import.module
+  imports.foldl (fun m i => process env i m) (({} : NameMap _).insert main imports)
+where
+  process (env) (i) (m) : NameMap (Array Name) :=
+    if m.contains i then
+      m
+    else
+      let imports := env.importsOf i
+      imports.foldr (fun i m => process env i m) (m.insert i imports)
+
+end Lean.Environment
+
+/--
+Return the redundant imports (i.e. those transitively implied by another import)
+amongst a candidate list of imports.
+-/
+partial def findRedundantImports (env : Environment) (imports : Array Name) : Array Name :=
+  let run := visit env.importGraph imports
+  let (_, seen) := imports.foldl (fun ⟨v, s⟩ n => run v s n) ({}, {})
+  seen.toArray
+where
+  visit (Γ) (targets) (visited) (seen) (n) : NameSet × NameSet :=
+    if visited.contains n then
+      (visited, seen)
+    else
+      let imports := (Γ.find? n).getD #[]
+      let (visited', seen') := imports.foldl (fun ⟨v, s⟩ t => visit Γ targets v s t) (visited, seen)
+      (visited'.insert n,
+        imports.foldl (fun s t => if targets.contains t then s.insert t else s) seen')
+
+/--
+Return the redundant imports (i.e. those transitively implied by another import)
+of a specified module (or the current module if `none` is specified).
+-/
+def redundantImports (n? : Option Name := none) : CoreM (Array Name) := do
+  let env ← getEnv
+  let imports := env.importsOf (n?.getD ((← getEnv).header.mainModule))
+  return findRedundantImports env imports
+
+/--
+List the imports in this file which can be removed
+because they are transitively implied by another import.
+-/
+elab "#redundant_imports" : command => do
+  let redundant := (← Elab.Command.liftCoreM do redundantImports)
+  if redundant.isEmpty then
+    logInfo "No transitively redundant imports found."
+  else
+    logInfo <| "Found the following transitively redundant imports:\n" ++
+      m!"{Format.joinSep redundant.toList "\n"}"
+
+/--
+Return the names of the modules in which constants used in the current file were defined,
+with modules already transitively imported removed.
+
+Note that this will *not* account for tactics and syntax used in the file,
+so the results may not suffice as imports.
+-/
+def Lean.Environment.minimalRequiredModules (env : Environment) : Array Name :=
+  let required := env.requiredModules.toArray
+  let redundant := findRedundantImports env required
+  required.filter fun n => ¬ redundant.contains n
```

Comment:
making `redundant` a `NameSet` will reduce the time complexity of this function, right? (`O(n*log n)` instead of `O(n^2)` for the last line?)

Reply:
Good point. It was in fact already a `NameSet`, that I was pointlessly turning into an `Array` too early.